### PR TITLE
refactor(batch): Use futures-async-stream to implement SortMergeJoin executor

### PR DESCRIPTION
**Pull request by D2Lark**
Signed-off-by: d2lark <lichengamoy@gmail.com>

Implement SortMergeJoin using Executor2 trait.

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)
close https://github.com/singularity-data/risingwave/issues/1949

### CODE REVIEW

**Comment by TennyZhuang** [comment]
CI: https://github.com/singularity-data/risingwave/actions/runs/2243765265

File: src/workspace-hack/Cargo.toml [hunk 1]
```diff
@@ -42,6 +42,7 @@ parking_lot = { version = "0.12", features = ["arc_lock", "deadlock_detection"]
 parking_lot_core = { version = "0.9", default-features = false, features = ["backtrace", "deadlock_detection", "petgraph", "thread-id"] }
 petgraph = { version = "0.6", features = ["graphmap", "matrix_graph", "stable_graph"] }
 prometheus = { version = "0.13", features = ["libc", "process", "procfs", "protobuf"] }
+prost = { version = "0.9", features = ["prost-derive", "std"] }
```
Comment (BowenXiao1999) [inline src/workspace-hack/Cargo.toml:45]: Why Cargo.toml change?

**Review by BowenXiao1999 — COMMENTED** [review]
General LGTM

**Comment by TennyZhuang** [comment]
CI is running: https://github.com/singularity-data/risingwave/actions/runs/2267661117

**Comment by TennyZhuang** [comment]
Thanks!

**Pull request merged**